src/mpyl/projects/versioning.py

```diff
@@ -203,12 +203,42 @@ def upgrade(self, previous_dict: ordereddict) -> ordereddict:
     ProjectUpgraderOne31(),
     ProjectUpgraderOneFour15(),
     ProjectUpgraderOneFour18(),
-    ProjectUpgraderOneFour20(),
+    ProjectUpgraderOneSix7(),
 ]
 
 
+class ConfigUpgraderOne412(Upgrader):
+    target_version = "1.4.12"
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
+        if kubernetes := previous_dict.get("kubernetes"):
+            if rancher := kubernetes.get("rancher"):
+                if cluster := rancher.get("cluster"):
+                    envs = ["pr", "test", "acceptance", "production"]
+
+                    clusters = [cluster.get(env) for env in envs if cluster.get(env)]
+                    named_clusters = [{"name": c["clusterEnv"], **c} for c in clusters]
+
+                    kubernetes["clusters"] = named_clusters
+                    kubernetes.pop("rancher")
+
+                    kubernetes["defaultCluster"] = {
+                        env: next(
+                            (
+                                c["clusterEnv"]
+                                for c in named_clusters
+                                if c["name"] == env
+                            ),
+                            "MISSING",
```

Comment:
What idea did you have with this default value?

Reply:
that we just write MISSING so it's obviously not right

Reply:
I was thinking that it's a bit obscure maybe, why not throw an error instead? Because this will error out later anyway right?